Signal/src/view controllers/MessagesViewController.m

```diff
@@ -495,50 +492,166 @@ - (void)setNavigationTitle
+    switch (rightBarButtonItemCount) {
+        case 0:
+            rightBarButtonSize = 65;
+            break;
+        case 1:
+            rightBarButtonSize = 100;
+            break;
+        default:
+            OWSAssert(0);
+        case 2:
+            rightBarButtonSize = 145;
+            break;
+    }
+    CGFloat maxTitleViewWidth = kShortScreenDimension - rightBarButtonSize;
+    const CGFloat titleViewWidth = MIN(maxTitleViewWidth,
+                                       MAX(self.navigationBarTitleLabel.frame.size.width,
+                                           self.navigationBarSubtitleLabel.frame.size.width));
+    self.navigationBarTitleView.frame = CGRectMake(0, 0,
+                                                   titleViewWidth,
+                                                   self.navigationBarTitleLabel.frame.size.height +
+                                                   self.navigationBarSubtitleLabel.frame.size.height +
+                                                   kTitleVSpacing);
+    self.navigationBarTitleLabel.frame = CGRectMake(0,
+                                                    0,
+                                                    titleViewWidth,
+                                                    self.navigationBarTitleLabel.frame.size.height);
+    self.navigationBarSubtitleLabel.frame = CGRectMake(0,
+                                                       self.navigationBarTitleView.frame.size.height - self.navigationBarSubtitleLabel.frame.size.height,
+                                                       titleViewWidth,
+                                                       self.navigationBarSubtitleLabel.frame.size.height);
+    
+    self.navigationItem.leftBarButtonItems = @[
+                                               backItem,
+                                               [[UIBarButtonItem alloc] initWithCustomView:self.navigationBarTitleView],
+                                               ];
+    
     if (self.userLeftGroup) {
         self.navigationItem.rightBarButtonItems = @[];
         return;
     }
 
+    const CGFloat kBarButtonSize = 44;
     NSMutableArray<UIBarButtonItem *> *barButtons = [NSMutableArray new];
     if ([self canCall]) {
-        UIBarButtonItem *callButton = [[UIBarButtonItem alloc] initWithImage:[UIImage imageNamed:@"btnPhone--white"]
-                                                                       style:UIBarButtonItemStylePlain
-                                                                      target:self
-                                                                      action:@selector(callAction)];
+        // We use UIButtons with [UIBarButtonItem initWithCustomView:...] instead of
+        // UIBarButtonItem in order to ensure that these buttons are spaced tightly.
+        // The contents of the navigation bar are cramped in this view.
+        UIButton *callButton = [UIButton buttonWithType:UIButtonTypeCustom];
```

Comment:
nit: They feel a little *too* tight on my iphone 7.

i.e. compare to other apps with two right bar buttons.